docs/spec.md

```diff
@@ -44,6 +44,56 @@ If a method called with incorrect types the behaviour is unspecified.
 Also if a method accepts a function it must only apply the function in accordance with
 the type signature (i.e. provide the correct number of arguments of the correct types).
 
+
+## Parametricity
+
+All methods' implementations should only use type information about arguments that is known from the
+method's type signature. It's not allowed to inspect arguments or values that they produce to get
+more information about their types. In other words they should be [parametrically polymorphic](https://en.wikipedia.org/wiki/Parametric_polymorphism).
+
+For example let's take a look at Functor's `map` signature:
+
+```
+map :: Functor f => Type f ~> (a → b, f a) → f b
+```
+
+There are three type variables in it: `f`, `a`, and `b`. Also we have some restrictions:
+
+  1. `Functor f` says that `f a` is a value of a Functor.
+  2. `Type f ~>` means that we're writing implementation of `map`
+     for a [type dictionary](#type) `Type f`. In this case we know what specific
+     types `Type f` works with, so we know everything about `f`.
+
+We don't have any restrictions for types `a` and `b`, so we don't know anything about them.
+And we're not allowed to inspect them.
+
+Here's an example of a correct `map` implementation that works with a simple wrapper `{x: ...}`:
+
+```js
+X.map = (f, v) => {
+  const a = v.x
+  const b = f(a)
+  return {x: b}
+}
+```
+
+And here's one that violates parametricity requirement although fits into type signature otherwise:
```

Comment:
Here's a common example of a parametricity violation (and the havoc it causes)

```javascript

// Assume `goodMap : (a -> b, Maybe a) -> Maybe b`

// Same type, but devious!
// badMap : (a -> b, Maybe a) -> Maybe b
function badMap(f, m) {
  if (m.value == 0) {
    return Maybe.Nothing
  } else {
    return goodMap(f, m)
  }
}
```

This is obviously cheating—it relies on (heterogenous) equality being defined for all values in Javascript—but nothing about the bare type variable `a` reflects that idea. If parametricity holds then we know certain so-called "free theorems" about the behavior of `badMap`, but since we've violated parametricity those will not hold.

Here's the free laws we can expect: http://www-ps.iai.uni-bonn.de/cgi-bin/free-theorems-webui.cgi where you type in `(a -> b) -> (Maybe a -> Maybe b)` and click "hide type instantiations" and then look at the bottom where "Reducing all permissable relation variables to functions" has happened

```haskell
-- our function is named `f`
(forall x . h (p x) = q (g x)) ==> forall x . fmap h (f p x) = f q (fmap g y) -- or
h . p = q . g ==> fmap h . f p = f q . fmap g
```

or, any such function should commute with `fmap` so long as the lifted functions commute as well. Let `h = p = q = g` and we'll see that any such function should commute with `fmap` generally.

    fmap g . f g = f g . fmap g

This obvious is violated by badmap since it has special behavior which blows up on hitting a particular value and that blow-up won't commute.